proposals/0021-vk-coop-matrix.md

```diff
@@ -0,0 +1,346 @@
+file "vk/util/arithmetic_selector.h". It can be used to generate inline Spir-V
+arithmetic instructions with the correct opcode for the type.
+
+The interface for `vk::khr::CooperativeMatrix` uses enums that are defined in
+the Spir-V specification. The required enums will be defined in "vk/spirv.h".
+
+## Detailed design
+
+### `vk::khr::CooperativeMatrix`
+
+The `CooperativeMatrix` class will defined in a header file as a wrapper around
+a `vk::SpirvType` that will expand to the appropriate SPIR-V type. This class
+will have the following interface.
+
+```c++
+
+// The base cooperative matrix class. The template arguments correspond to the
+// operands in the OpTypeCooperativeMatrixKHR instruction.
+template <typename ComponentType, Scope scope, uint rows, uint columns,
+          CooperativeMatrixUse use>
+class CooperativeMatrix {
+  template <class NewComponentType>
+  CooperativeMatrix<NewComponentType, scope, rows, columns, use> cast();
+
+  // Apply OpSNegate or OFNegate, depending on ComponentType, in a element by
+  // element manner.
+  CooperativeMatrix negate();
+
+  // Apply OpIAdd or OFAdd, depending on ComponentType, in a element by element
+  // manner.
+  CooperativeMatrix operator+(CooperativeMatrix other);
+
+  // Apply OpISub or OFSub, depending on ComponentType, in a element by element
+  // manner.
+  CooperativeMatrix operator-(CooperativeMatrix other);
+
+  // Apply OpIMul or OFMul, depending on ComponentType, in a element by element
+  // manner.
+  CooperativeMatrix operator*(CooperativeMatrix other);
+
+  // Apply OpSDiv, OpUDiv or OFDiv, depending on ComponentType, in a element by
+  // element manner.
+  CooperativeMatrix operator/(CooperativeMatrix other);
+
+  // Apply OpMatrixTimesScalar in a element by element manner.
+  CooperativeMatrix operator*(ComponentType scalar);
+
+  // Note: The SPIR-V specification allows the load and store function to access
+  // workgroup data (groupshared in HLSL). This call does not accept group
+  // shared data because HLSL cannot pass the groupshared array by reference. It
+  // is impossible to implement a function that will load and store from group
+  // sharded data. This can be revisited when references are added to HLSL.
+
+  // Store the cooperative matrix using OpCooperativeMatrixStoreKHR to data[i]
+  // using memory layout RowMajorKHR.￼
+  template <class Type>
+  void StoreRowMajor(RWStructuredBuffer<Type> data, uint32_t index);
+
+  // Store the cooperative matrix using OpCooperativeMatrixStoreKHR to data[i]
+  // using memory layout ColumnMajorKHR.￼
+  template <class Type>
+  void StoreColumnMajor(RWStructuredBuffer<Type> data, uint32_t index);
+
+  // Store the cooperative matrix using OpCooperativeMatrixStoreKHR to data[i]
+  // using memory layout RowMajorKHR and the given stride and memory access
+  // mask.
+  template <class Type>
+  void StoreRowMajor(RWStructuredBuffer<Type> data, uint32_t index,
```

Comment:
It is not necessary. This was just my personal preference for functions with fewer parameters. If there will be more layouts, it might be better to make it a parameter.